file.h

```diff
@@ -1455,6 +1367,122 @@ std::uint64_t basic_file_base<File>::tell() const
 #endif
 }
 
+template <typename File>
+std::vector<std::byte> basic_file_base<File>::read_nonzero(std::size_t bytes_to_read) const
+{
+    std::vector<std::byte> data(bytes_to_read);
+    auto actually_read = read({ data.data(), bytes_to_read });
+    data.resize(actually_read);
+
+    // Consider omitting
+    data.shrink_to_fit();
+
+    return data;
+}
+
+template <typename File>
+std::vector<std::byte> basic_file_base<File>::read_zero() const
+{
+    // 0 means that the caller is requesting to read all available
+    // try to find out how many bytes can be read
+
+    std::optional<std::size_t> bytes_can_be_read = try_get_bytes_can_be_read();
+
+    if (!bytes_can_be_read.has_value()) {
+        // Cannot determine apriori how many bytes can be read.
+        return read_and_resize_to_end();
+    }
+
+    auto bytes_to_read = bytes_can_be_read.value();
+    if (bytes_to_read == 0) {
+        // Nothing to read.
+        return {};
+    }
+    return read_nonzero(bytes_to_read);
+}
+
+template <typename File>
+std::optional<std::size_t> basic_file_base<File>::try_get_bytes_can_be_read() const
+{
+    auto file_size = this->size();
+    if (file_size == 0) {
+        // Cannot determine file size
+        return std::nullopt;
+    }
+    auto current_offset = tell();
+    if (tell() >= file_size) {
```

Comment:
Calling tell twice seems risky, in any case, can optimize the two `tell` calls into one.

Reply:
Of course. My bad.